### ISSUE TITLE
`==`(x,y) is true when only partials(y) have NaNs

### ISSUE THREAD

**Issue by magerton** [JuliaDiff/ForwardDiff.jl#600]
When `x` and `y` are duals and have equal values, but the partials differ in that some are NaNs, the statement `x==y` evaluates to true, even though partials(x) == partials(y) is not true. I'm guessing that this is not the intended behavior, but I'm not sure where the fix would go. The problem doesn't seem to hold for `Partials`

```julia
using ForwardDiff: Dual, Partials, partials
Dual(0.0, (1.0, 0.0)) == Dual(0.0, (1.0, NaN)) # true... should be false?
Dual(0.0, (1.0, 0.0)) == Dual(0.0, (NaN, NaN)) # true... should be false?
Dual(0.0, (1.0, 0.0)) == Dual(NaN, (1.0, 0.0)) # false

Partials((0.0, 1.0)) == Partials((0.0, 1.0)) # true
Partials((0.0, 1.0)) == Partials((NaN, 1.0)) # false

partials(Dual(0.0, (1.0, 0.0))) == partials(Dual(0.0, (1.0, 0.0))) # true
partials(Dual(0.0, (1.0, 0.0))) == partials(Dual(0.0, (1.0, NaN))) # false
partials(Dual(0.0, (1.0, 0.0))) == partials(Dual(0.0, (NaN, NaN))) # false
```

**Comment by magerton** [comment]
Hi @mcabbott , thanks for the quick response.

I'm not sure what you mean by the "tests are as you desire." What was strange (and **undesirable** to me) was that `x==y` evaluated to `true` even though the dual parts (`Partials`) were not equal. If this is fixed in `master`, we should close this issue.

To answer your question, I noticed this when testing a backslash (`\`) method I'm writing to handle a sparse matrix of duals (a derivative of #363). For my small test case, the dense version of `\` returned a vector of duals with partials equal to zero, while the sparse version of the function returned NaNs. The fact that `Msparse\b == Mdense\b` when the partials were different (eg, one had `NaN`s) is problematic for verifying that my function is returning the same result for dense and sparse matrices.

https://github.com/magerton/ForwardDiffMatrixTools.jl/blob/eb56f763abc6fc87ea4f58b6e29fdec5f9f13ae6/src/ForwardDiffMatrixTools.jl#L66-L106

**Comment by magerton** [comment]
In case helpful, I'm using
```julia
[f6369f11] ForwardDiff v0.10.32
```

**Comment by mcabbott** [comment]
Ok, good. So your test `Msparse\b == Mdense\b` will (soon) detect the presence of NaNs, although presumably the final calculation once it all works will not contain them. 

v0.10.32 is without this PR. My results above are with `] add ForwardDiff#master`, which includes it.

**Comment by magerton** [comment]
Perfect. Thank you!! 

Closed by #481
